### PR TITLE
Better sorting of search results (fixes #100 and #603)

### LINKED ISSUE #100: Implement fuzzy search
Devdocs.io has a feature called fuzzy search, for example "pmsg" would autocomplete to "window.postMessage".

I would like to see this implemented in Zeal.

Thanks for this project, really good !

### PR DESCRIPTION
**Pull request by jkozera**
(Accidentally, adds fuzzy search too.)

Uses an O(m+n) algorithm based on https://github.com/bevacqua/fuzzysearch - should be faster than the one initially proposed in PR #281.

Screenshots:
screenshot from 2016-10-07 15-28-12

screenshot from 2016-10-07 16-51-26

### CODE REVIEW

File: src/libs/registry/qsqlitelite.h [hunk 1]
```diff
@@ -0,0 +1,31 @@
+#ifndef QSQLITELITE_H
```
Comment (trollixx) [inline src/libs/registry/qsqlitelite.h]: This file should not be here :)

File: src/libs/util/sqlitedatabase.cpp [hunk 2]
```diff
@@ -159,3 +159,8 @@ void SQLiteDatabase::updateLastError()
 
     m_lastError = QString(reinterpret_cast<const QChar *>(sqlite3_errmsg16(m_db)));
 }
+
+sqlite3 *SQLiteDatabase::handle()
```
Comment (trollixx) [inline src/libs/util/sqlitedatabase.cpp]: As mentioned above, I like the idea of not exposing SQLite API directly.

File: src/libs/registry/docset.cpp [hunk 3]
```diff
@@ -607,3 +612,178 @@ QString Docset::parseSymbolType(const QString &str)
...
+    if (match1 == -1) { // no match
+        // simply return best=0 below
+    } else if (needleLen == match1Len) { // exact match
+        best = scoreExact(match1, match1Len, haystack, haystackLen);
+    } else {
+        best = scoreFuzzy(match1, match1Len, haystack);
+
+        int indexOfLastDot = -1;
+        for (int i = 0; haystack[i] != 0; ++i) {
+            if (haystack[i] == '.') indexOfLastDot = i;
```
Comment (trollixx) [inline src/libs/registry/docset.cpp]: No one-liners.

File: src/libs/registry/docset.cpp [hunk 4]
```diff
@@ -607,3 +612,178 @@ QString Docset::parseSymbolType(const QString &str)
...
+        }
+
+        if (indexOfLastDot != -1) {
+            int match2 = -1, match2Len;
+            matchFuzzy(
```
Comment (trollixx) [inline src/libs/registry/docset.cpp]: Just break arguments wherever needed, don't leave the whole line empty.

File: src/libs/registry/docset.cpp [hunk 5]
```diff
@@ -607,3 +612,178 @@ QString Docset::parseSymbolType(const QString &str)
...
+                needleLen, needle, haystackLen - (indexOfLastDot + 1), haystack + indexOfLastDot + 1,
+                &match2, &match2Len
+            );
+            if (match2 != -1) {
+                best = qMax(
+                    best, scoreFuzzy(match2, match2Len, haystack + indexOfLastDot + 1)
```
Comment (trollixx) [inline src/libs/registry/docset.cpp]: Same as above.

File: src/libs/registry/docset.cpp [hunk 6]
```diff
@@ -607,3 +612,178 @@ QString Docset::parseSymbolType(const QString &str)
...
+            haystack[j++] = c;
+        }
+    }
+
+    int best = 0, match1 = -1, match1Len;
+
+    matchFuzzy(needleLen, needle, haystackLen, haystack, &match1, &match1Len);
+
+    if (match1 == -1) { // no match
+        // simply return best=0 below
```
Comment (trollixx) [inline src/libs/registry/docset.cpp]: Return ASAP.

File: src/libs/registry/docset.cpp [hunk 7]
```diff
@@ -607,3 +612,178 @@ QString Docset::parseSymbolType(const QString &str)
...
+            );
+            if (match2 != -1) {
+                best = qMax(
+                    best, scoreFuzzy(match2, match2Len, haystack + indexOfLastDot + 1)
+                );
+            }
+        }
+    }
+
+    delete[] needle;
```
Comment (trollixx) [inline src/libs/registry/docset.cpp]: Use [`QScopedArrayPointer`](http://doc.qt.io/qt-5/qscopedarraypointer.html) instead.

File: src/libs/registry/docset.cpp [hunk 8]
```diff
@@ -607,3 +612,178 @@ QString Docset::parseSymbolType(const QString &str)
...
+        } else {
+            return qMax(1, 34 - matchLen);
+        }
+    }
+}
+
+static int scoreFunc(const char *needleOrig, const char* haystackOrig)
+{
+    int haystackLen = 0, needleLen = 0;
+    while (haystackOrig[++haystackLen] != 0);
```
Comment (trollixx) [inline src/libs/registry/docset.cpp]: This is hard to understand, please replace with a loop with `++haystackLen` inside.

File: src/libs/registry/docset.cpp [hunk 9]
```diff
@@ -607,3 +612,178 @@ QString Docset::parseSymbolType(const QString &str)
...
+    while (needleOrig[++needleLen] != 0);
```
Comment (trollixx) [inline src/libs/registry/docset.cpp]: Same as above.

File: src/libs/registry/docset.cpp [hunk 10]
```diff
@@ -121,6 +125,10 @@ Docset::Docset(const QString &path) :
         return;
     }
 
+    sqlite3_create_function(
```
Comment (trollixx) [inline src/libs/registry/docset.cpp]: Please put arguments on the same line.

File: src/libs/registry/docset.cpp [hunk 11]
```diff
@@ -36,6 +36,10 @@
 #include <QRegularExpression>
 #include <QVariant>
 
+#include <sqlite3.h>
+
+static void scoreFunc(sqlite3_context *context, int /* argc */, sqlite3_value **argv);
```
Comment (trollixx) [inline src/libs/registry/docset.cpp]: No need to comment parameters in a function declaration.

File: src/libs/registry/docset.cpp [hunk 12]
```diff
@@ -607,3 +617,183 @@ QString Docset::parseSymbolType(const QString &str)
 
     return aliases.value(str, str);
 }
+
+// ported from DevDocs' searcher.coffee:
+// (https://github.com/Thibaut/devdocs/blob/50f583246d5fbd92be7b71a50bfa56cf4e239c14/assets/javascripts/app/searcher.coffee#L91)
+static void matchFuzzy(
```
Comment (trollixx) [inline src/libs/registry/docset.cpp]: Just put parameters on the same line, and break to fit the per line character limit.

File: src/libs/registry/docset.cpp [hunk 13]
```diff
@@ -607,3 +617,183 @@ QString Docset::parseSymbolType(const QString &str)
...
+    int nLen, const unsigned char *needle, int hLen, const unsigned char *haystack,
+    int *start, int *len
+) {
+    int j = 0, groups = 0;
```
Comment (trollixx) [inline src/libs/registry/docset.cpp]: Should be on different lines.

File: src/libs/registry/docset.cpp [hunk 14]
```diff
@@ -607,3 +617,183 @@ QString Docset::parseSymbolType(const QString &str)
...
+    for (int i = 0; i < nLen; ++i) {
+        bool found = false, first = true;
```
Comment (trollixx) [inline src/libs/registry/docset.cpp]: Two lines here too please.

File: src/libs/registry/docset.cpp [hunk 15]
```diff
@@ -607,3 +617,183 @@ QString Docset::parseSymbolType(const QString &str)
...
+        int distance = 0;
+        while (j < hLen) {
+            bool match = needle[i] == haystack[j++];
+            if (match) {
```
Comment (trollixx) [inline src/libs/registry/docset.cpp]: `match` is not used anywhere else, just use condition in `if`. Also simply `++j` one line above.

File: src/libs/registry/docset.cpp [hunk 16]
```diff
@@ -607,3 +617,183 @@ QString Docset::parseSymbolType(const QString &str)
...
+                if (*start == -1) *start = j - 1;  // first matched char
```
Comment (trollixx) [inline src/libs/registry/docset.cpp]: Needs a line break.

File: src/libs/registry/docset.cpp [hunk 17]
```diff
@@ -607,3 +617,183 @@ QString Docset::parseSymbolType(const QString &str)
...
+            } else {
+                // optimizations to reduce returned number of results
+                // (search was returning too many irrelevant results with large docsets)
+                if (first) {
+                    groups += 1;
+                    if (groups > 3)  // optimization #1: too many mismatches
+                        break;
+                    first = false;
+                }
+                if (i != 0) {
```
Comment (trollixx) [inline src/libs/registry/docset.cpp]: A line break

File: src/libs/registry/docset.cpp [hunk 18]
```diff
@@ -607,3 +617,183 @@ QString Docset::parseSymbolType(const QString &str)
...
+            // beginning of the string or is preceded by a dot.
+            return 0;
+        } else {
+            // (1) Remove one point for each unmatched character up to
+            //     the nearest preceding dot or the beginning of the
+            //     string.
+            // (2) Remove one point for each unmatched character
+            //     following the query.
+            int i = matchIndex - 2;
+            while (i >= 0 && value[i] != DOT) --i;
```
Comment (trollixx) [inline src/libs/registry/docset.cpp]: Line break.

File: src/libs/registry/docset.cpp [hunk 19]
```diff
@@ -607,3 +617,183 @@ QString Docset::parseSymbolType(const QString &str)
...
+            score -= (matchIndex - i) +                     // (1)
```
Comment (trollixx) [inline src/libs/registry/docset.cpp]: When doing a line break, always put an operator on the next line.

File: src/libs/registry/docset.cpp [hunk 20]
```diff
@@ -607,3 +617,183 @@ QString Docset::parseSymbolType(const QString &str)
...
+            // score between 33..66, if the match is at the end of the string
+            return qMax(33, 67 - matchLen);
+        } else {
+            // score between 1..33 otherwise (match in the middle of the string)
+            return qMax(1, 34 - matchLen);
+        }
+    }
+}
+
+static void scoreFunc(sqlite3_context *context, int /* argc */, sqlite3_value **argv) {
```
Comment (trollixx) [inline src/libs/registry/docset.cpp]: `Q_UNUSED` macro is the preferable way of preventing comiler warning when some of function parameters are not in use.

File: src/libs/registry/docset.cpp [hunk 21]
```diff
@@ -607,3 +617,183 @@ QString Docset::parseSymbolType(const QString &str)
...
+        ++needleLen;
+    QScopedArrayPointer<unsigned char> needle(new unsigned char[needleLen + 1]);
+    QScopedArrayPointer<unsigned char> haystack(new unsigned char[haystackLen + 1]);
+    for (int i = 0; i < needleLen + 1; ++i) {
+        char c = needleOrig[i];
+        if (c >= 'A' && c <= 'Z')
+            c += 32;
+        needle[i] = c;
+    }
+    for (int i = 0, j = 0; i < haystackLen + 1; ++i) {
```
Comment (trollixx) [inline src/libs/registry/docset.cpp]: A line break above.

File: src/libs/registry/docset.cpp [hunk 22]
```diff
@@ -607,3 +617,183 @@ QString Docset::parseSymbolType(const QString &str)
...
+        char c = haystackOrig[i];
+        if ((i > 0 && haystackOrig[i - 1] == ':' && c == ':') // C++ (::)
+                || c == '/' || c == '_' || c == ' ') { // Go, some Guides
+            haystack[j++] = '.';
```
Comment (trollixx) [inline src/libs/registry/docset.cpp]: Just do `++j` at the end of the loop block.

File: src/libs/registry/docset.cpp [hunk 23]
```diff
@@ -607,3 +617,183 @@ QString Docset::parseSymbolType(const QString &str)
...
+        } else {
+            if (c >= 'A' && c <= 'Z')
+                c += 32;
+            haystack[j++] = c;
```
Comment (trollixx) [inline src/libs/registry/docset.cpp]: Same as above.

File: src/libs/util/sqlitedatabase.cpp [hunk 24]
```diff
@@ -79,9 +79,14 @@ bool SQLiteDatabase::execute(const QString &queryStr)
 
     m_lastError.clear();
 
+    sqlite3_mutex_enter(sqlite3_db_mutex(m_db));
     const void *pzTail = nullptr;
-    if (sqlite3_prepare16_v2(m_db, queryStr.constData(), (queryStr.size() + 1) * sizeof(QChar),
-                             &m_stmt, &pzTail) != SQLITE_OK) {
+    const int res = sqlite3_prepare16_v2(
```
Comment (trollixx) [inline src/libs/util/sqlitedatabase.cpp]: Parameters should start on this line.

File: src/libs/registry/docset.cpp (hunk 15 above)
Comment (jkozera) [inline src/libs/registry/docset.cpp]: j++ is actually necessary here, as opposed to ++j, which would omit the first character when comparing.

**Pull request merged**